src/linearAlgebra/Vector.h

```diff
@@ -87,280 +129,254 @@ namespace dftefe
-       * @return \f$ l_{\inf} \f$  norm of the vector as double type
+       * @return \f$ l_{\inf} \f$  norm of the vector
        */
-      virtual double
-      lInfNorm() const = 0;
-
-      virtual void
-      updateGhostValues(const size_type communicationChannel = 0) = 0;
+      double
+      lInfNorm() const;
 
-      virtual void
-      accumulateAddLocallyOwned(const size_type communicationChannel = 0) = 0;
-
-      virtual void
-      updateGhostValuesBegin(const size_type communicationChannel = 0) = 0;
+      void
+      updateGhostValues(const size_type communicationChannel = 0);
 
-      virtual void
-      updateGhostValuesEnd() = 0;
+      void
+      accumulateAddLocallyOwned(const size_type communicationChannel = 0);
 
-      virtual void
-      accumulateAddLocallyOwnedBegin(
-        const size_type communicationChannel = 0) = 0;
+      void
+      updateGhostValuesBegin(const size_type communicationChannel = 0);
 
-      virtual void
-      accumulateAddLocallyOwnedEnd() = 0;
+      void
+      updateGhostValuesEnd();
 
-    protected:
-      /**
-       * @brief Constructor
-       *
-       * @param[in] storage reference to unique_ptr to Vector::Storage (i.e.,
-       * MemoryStorage) from which the Vector to transfer ownership.
-       * @param[in] globalSize global size of the vector (i.e., the number of
-       * unique indices across all processors).
-       * @param[in] locallyOwnedSize size of the part of the vector for which
-       * the current processor is the sole owner (see top for explanation). For
-       * a SerialVector, the locallyOwnedSize and the the globalSize are the
-       * same.
-       * @param[in] ghostSize size of the part of the vector that is owned by
-       * the other processors but required by the current processor. For a
-       * SerialVector, the ghostSize is 0.
-       * @param[in] pointer to LinAlgOpContext object which stores hardware
-       * specific handles for linear algebra operations on HOST/DEVICE.
-       *
-       * @note Since we are passing the ownership of the input storage to the Vector, the
-       * storage will point to NULL after a call to this Constructor. Accessing
-       * the input storage pointer will lead to undefined behavior.
-       */
-      Vector(std::unique_ptr<Storage> &    storage,
-             const global_size_type        globalSize,
-             const size_type               locallyOwnedSize,
-             const size_type               ghostSize,
-             LinAlgOpContext<memorySpace> *linAlgOpContext);
+      void
+      accumulateAddLocallyOwnedBegin(const size_type communicationChannel = 0);
 
-      /**
-       * @brief Default Constructor
-       */
-      Vector();
+      void
+      accumulateAddLocallyOwnedEnd();
 
-    protected:
+    private:
       std::unique_ptr<Storage>      d_storage;
       LinAlgOpContext<memorySpace> *d_linAlgOpContext;
       VectorAttributes              d_vectorAttributes;
```

Comment:
Should we remove d_vectorAttributes as we discussed earlier?

Reply:
It is useful in cases where we are operating two or more Vectors and would require compatibility (i.e., all must be serial or all must be distributed) for the operation to make sense.